File: JavaScript/974-Subarray Sums Divisible by K.js
/**
 * @param {number[]} A
 * @param {number} K
 * @return {number}
 */
var subarraysDivByK = function(A, K) {
    let preSum = [0];
    let hmap = new Map();
    hmap.set(0, 1);
    
    for (let a of A){
        let curr = preSum[preSum.length-1] + a  //note that (-1%5) = -1 in JavaScript, therefore we use ((-1%5) + 5)%5 instead
        let num = ((curr%K)+K)%K
        preSum.push(num);
        hmap.set(num,  (hmap.get(num)||0)+1 )
    }
    
    let res = 0;
    for (let val of hmap.values()){
        res += val*(val-1)/2;
    }
    return res;
};

File: C++/974-Subarray Sums Divisible by K.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        vector<int> cnt(k, 0);
        cnt[0] = 1;
        int curr = 0;
        int res = 0;

        for (int num:nums) {
            curr = (curr + num%k + k)%k;   // a = (a/b)*b + a%b, a%b = a - (a/b)*b, so sometimes a%b gives negative results.
            res += cnt[curr];
            cnt[curr]++;
        }

        return res;
    }
};